freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,371 @@
+        """
+        with open(self.pair_dictionary_path, "w") as fp:
+            json.dump(self.pair_dict, fp, default=self.np_encoder)
+
+    def save_follower_dict_to_disk(self):
+        """
+        Save follower dictionary to disk (used by strategy for persistent prediction targets)
+        """
+        with open(self.follower_dict_path, "w") as fp:
+            json.dump(self.follower_dict, fp, default=self.np_encoder)
+
+    def create_follower_dict(self):
+        """
+        Create or dictionary for each follower to maintain unique persistent prediction targets
+        """
+
+        whitelist_pairs = self.config.get("exchange", {}).get("pair_whitelist")
+
+        exists = self.follower_dict_path.is_file()
+
+        if exists:
+            logger.info("Found an existing follower dictionary")
+
+        for pair in whitelist_pairs:
+            self.follower_dict[pair] = {}
+
+        with open(self.follower_dict_path, "w") as fp:
+            json.dump(self.follower_dict, fp, default=self.np_encoder)
+
+    def np_encoder(self, object):
+        if isinstance(object, np.generic):
+            return object.item()
+
+    def get_pair_dict_info(self, pair: str) -> Tuple[str, int, bool, bool]:
+        """
+        Locate and load existing model metadata from persistent storage. If not located,
+        create a new one and append the current pair to it and prepare it for its first
+        training
+        :param pair: str: pair to lookup
+        :return:
+            model_filename: str = unique filename used for loading persistent objects from disk
+            trained_timestamp: int = the last time the coin was trained
+            coin_first: bool = If the coin is fresh without metadata
+            return_null_array: bool = Follower could not find pair metadata
+        """
+        pair_in_dict = self.pair_dict.get(pair)
+        data_path_set = self.pair_dict.get(pair, {}).get("data_path", None)
+        return_null_array = False
+
+        if pair_in_dict:
+            model_filename = self.pair_dict[pair]["model_filename"]
+            trained_timestamp = self.pair_dict[pair]["trained_timestamp"]
+            coin_first = self.pair_dict[pair]["first"]
+        elif not self.follow_mode:
+            self.pair_dict[pair] = {}
+            model_filename = self.pair_dict[pair]["model_filename"] = ""
+            coin_first = self.pair_dict[pair]["first"] = True
+            trained_timestamp = self.pair_dict[pair]["trained_timestamp"] = 0
+            self.pair_dict[pair]["priority"] = len(self.pair_dict)
+
+        if not data_path_set and self.follow_mode:
+            logger.warning(
+                f"Follower could not find current pair {pair} in "
+                f"pair_dictionary at path {self.full_path}, sending null values "
+                "back to strategy."
+            )
+            return_null_array = True
+
+        return model_filename, trained_timestamp, coin_first, return_null_array
```

Comment:
coin_first can be unbound (it's not set if `pair_in_dict` and `follow_mode` are both false). This can lead to very subtle crashes of "used before initialized" errors.

Reply:
coin_first can be unbound (it's not set if `pair_in_dict` and `follow_mode` are both false). This can lead to very subtle crashes of "used before initialized" errors.